CruGlobal/mobile-content-api
# [GT-643] Global activity analytics API

Ticket: https://jira.cru.org/browse/GT-643

- [x] Create the request for access token (default expiration 24 hours) in exchange for JWT
- [x] Call Adobe Analytics through HTTP to get the data, use access token, endpoint is configurable
- [x] Create the table to keep necessary data
- [x] Create API controller exposes the data
- [x] Fill in the table with data coming from Adobe Analytics API
- [x] Trigger counters update

**Questions:**

- What is the format for access token request in exchange for JWT token?
Answer: 
```
curl -X POST -H "Cache-Control: no-cache" -H "Content-Type: multipart/form-data" -F "client_id=<ADOBE_ANALYTICS_CLIENT_ID>" -F "client_secret=<ADOBE_ANALYTICS_CLIENT_SECRET>" -F "jwt_token=<ADOBE_ANALYTICS_JWT_TOKEN>" "<ADOBE_ANALYTICS_EXCHANGE_JWT_URL>"
```
Expect Response `status` 200 with `body`:
(note that the `expires_in` value will not be correct)
```
{
    "token_type": "bearer",
    "access_token": "lkjasdfkjlhasdflkjhasdflkjhasdfjkhaslkjhalskjdfh",
    "expires_in": 86399995
}
```

- What's the correct header for requests to Adobe Analytics? Is the following correct one?

Answer:
```
Authorization: Bearer <access token in response from request to ADOBE_ANALYTICS_EXCHANGE_JWT_URL>
x-api-key: <ADOBE_ANALYTICS_CLIENT_ID>
x-proxy-global-company-id: <ADOBE_ANALYTICS_COMPANY_ID>
```
- Request format: is the `Client ID` in the header correct?

Answer: No Client ID necessary.

- Response format: 

```
    "rows": [
        {
            "itemId": "1173709922",
            "value": "Some Country",
            "data": [
                123,
                123,
                123,
                218
            ]
        },
```
what the numbers in "data" array mean? `users countries launches gospel_presentations`? is that correct order?
Answer: 1 Unique Visitors, 2 Launches, 3 Presenting the Gospel, 4 Total Countries (row count)
3. Response format: can we take counters from this piece of data rather than reducing manually?

Answer: Yes.

```
    "summaryData": {
        "filteredTotals": [
            238326,
            966442,
            43834,
            218
        ],
        "totals": [
            238326,
            966442,
            43834,
            218
        ],
        "col-max": [
            42828,
            165291,
            11977,
            218
        ],
        "col-min": [
            1,
            1,
            1,
            218
        ]
    }
```

### Testing instructions

1. Define the following env vars:
```
    * ADOBE_ANALYTICS_REPORT_URL
    * ADOBE_ANALYTICS_COMPANY_ID
    * ADOBE_ANALYTICS_CLIENT_ID
    * ADOBE_ANALYTICS_JWT_TOKEN
    * ADOBE_ANALYTICS_CLIENT_SECRET
    * ADOBE_ANALYTICS_EXCHANGE_JWT_URL
```
2. Run rails server: `rails s`
3. Check the endpoint: `curl localhost:3000//analytics/global`
